### PR TITLE
Allow non-alpha-numeric characters in GitHub usernames #340

### LINKED ISSUE #340: Allow non-alpha-numeric characters in GitHub usernames
Error message:
```
12:41:22 - CS2103-AY1819S1-W10-2_main_master has no authors specified, using all authors by default.
Exception in thread "main" java.lang.IllegalArgumentException: The provided Git ID, Aiden Low (Yew Woei), contains illegal characters.
        at reposense.model.Author.<init>(Author.java:35)
        at reposense.git.GitShortlog.lambda$extractAuthorsFromLog$0(GitShortlog.java:24)
        at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
        at java.util.Spliterators$ArraySpliterator.forEachRemaining(Unknown Source)
        at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source)
        at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
        at java.util.stream.ReferencePipeline.collect(Unknown Source)
        at reposense.git.GitShortlog.extractAuthorsFromLog(GitShortlog.java:26)
        at reposense.report.ReportGenerator.updateAuthorList(ReportGenerator.java:112)
        at reposense.report.ReportGenerator.generateReposReport(ReportGenerator.java:66)
        at reposense.RepoSense.main(RepoSense.java:51)

```

### PR DESCRIPTION
**Pull request by yamidark**
Fixes #340 
```
The Author class restricts the github id and author names to only contain
alphanumeric characters, plus some other symbols, for security purposes as
this values are used inside the OS terminal.

As we are allowed to change out git author name to be anything as it
does not have any restrictions, having this restrictions for our Author
class will prevent some users from using our app to analyze their repo
if the author name they wish to use contain non alphanumeric characters.

Also, special non-UTF-8 characters may also be used for the author name,
which may cause MalformedInputException[1] when trying to read the
author-config.csv file.

As the --author param field of the git log command uses regex pattern
to perform the author filtration, we can swap out any of these computer
symbols which may have different uses in the OS terminal into a . period
instead to indicate any character instead, allowing us to still get the correct
author filter without conflicting with the OS terminal. However, this may
cause the accuracy of the analysis to be reduced if there are multiple authors
with very similar names in the repo.

Let's
- remove validation of github id and names in the Author class
- add a method in StringUtil to swap computer symbols in the author
name to the . period
- add a note in the UserGuide warning users who have author names
that are very similar and have symbols
- update the CsvParser to use the java.io BufferedReader constructor for
reading of the csv files instead.

[1]: Similar issue when reading files previously resolved
https://github.com/reposense/RepoSense/commit/1f222c0fd93fd8c5399a27c18aa03c03871f7cee
```

### CODE REVIEW

**Comment by yamidark** [comment]
@damithc Could you try testing this version on those repositories with the problems?

**Comment by damithc** [comment]
Try running against markdown-here repo
I'll send you the repo list of the current batch to try.

File: src/main/java/reposense/system/CommandRunner.java [hunk 1]
```diff
@@ -24,7 +24,7 @@
     private static final String AUTHOR_NAME_PATTERN = "^%s <.*>$";
     private static final String OR_OPERATOR_PATTERN = "\\|";
 
-    private static final Pattern SPECIAL_REGEX_CHARS = Pattern.compile("[{}()\\[\\].+*?^$|]");
+    private static final Pattern SPECIAL_SYMBOLS = Pattern.compile("[;:&/\\\\!<>{}%#\"\\-='()\\[\\].+*?^$|]");
```
Comment (eugenepeh) [inline src/main/java/reposense/system/CommandRunner.java]: can probably do like 
>!-/:-@[-\`

making use of range, but i'm okay with anything.

**Comment by yamidark** [comment]
>perhaps can add unit test for replaceSpecialSymbols

Its a private helper method though, don't think we want to make it public just for testing. 
Have also already added 1 test for it with the `gitlog` command as well.

**Comment by yamidark** [comment]
@eugenepeh Add the tests, and also added a note in UserGuide for users using names with computer symbols in them (will rebase after #344 is merged)

@damithc This version works for the markdown-here repo.
It also works for the current list of student repos, but the Chrome/Firefox browser isn't able to load all of it due to insufficient memory. I think this may be because we are finding all authors for the repo by not specifying a `author-config.csv`, so each of them will have the all the past contributors to AB4).

**Comment by damithc** [comment]
> It also works for the current list of student repos, but the Chrome/Firefox browser isn't able to load all of it due to insufficient memory. I think this may be because we are finding all authors for the repo by not specifying a `author-config.csv`, so each of them will have the all the past contributors to AB4).

Thank for checking. This is why I requested for #341 :-)

**Comment by yamidark** [comment]
>This is why I requested for #341 :-)

Will push a fix for it.

**Comment by yamidark** [comment]
@eugenepeh Rebased, for your review again

File: docs/UserGuide.md [hunk 2]
```diff
@@ -180,6 +180,8 @@ git config --global user.name "YOUR_AUTHOR_NAME”
 ```
 RepoSense expects the Git Author Name to be the same as author's GitHub username. If an author's `Git Author Name` is different from her `GitHub ID`, the `Git Author Name` needs to be specified in the standalone config file. If the author has more than one `Git Author Name`, multiple values can be entered too.
 
+> Note: As we use the OS terminal git command for analyse of repos, computer symbols (e.g. `"`, `!`, `/` etc.) inside the name will be replaced with a period `.` during analysis, which may reduce the accuracy of the analysis.
```
Comment (eugenepeh) [inline docs/UserGuide.md]: I think a bit too much info. Perhaps just a `Note: using symbols such as... in your author name may reduce the accuracy of the analysis if 2 names are approximately similar`. will do

Comment (yamidark) [reply docs/UserGuide.md]: Fixed

File: src/test/resources/StandaloneConfigTest/specialCharacterAuthor_config.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	[
+		{
+			"githubId": "‘Processed�‘Cooked�",
+			"displayName": "‘Processed�‘Cooked�",
+			"authorNames": ["‘Processed�‘Cooked�"]
+		},
+		{
+			"githubId": "Aiden Low (Yew Woei)",
+			"displayName": "Aiden Low (Yew Woei)",
+			"authorNames": ["Aiden Low (Yew Woei)"]
```
Comment (damithc) [inline src/test/resources/StandaloneConfigTest/specialCharacterAuthor_config.json]: Best not to use real names

**Comment by damithc** [comment]
Thanks for the fix @yamidark

**Pull request merged**